airbyte-oauth/src/main/java/io/airbyte/oauth/GoogleOAuthFlow.java

```diff
@@ -0,0 +1,220 @@
+    if (queryParams.containsKey("code")) {
+      final String code = (String) queryParams.get("code");
+      final JsonNode oAuthParamConfig = getSourceOAuthParamConfig(workspaceId, sourceDefinitionId);
+      if (oAuthParamConfig.has("client_id") && oAuthParamConfig.has("client_secret")) {
+        final String clientId = oAuthParamConfig.get("client_id").asText();
+        final String clientSecret = oAuthParamConfig.get("client_secret").asText();
+        return completeOAuthFlow(clientId, clientSecret, code, redirectUrl);
+      } else {
+        throw new IOException("Undefined parameter 'client_id' and 'client_secret' for Google OAuth Flow.");
+      }
+    } else {
+      throw new IOException("Undefined 'code' from consent redirected url.");
+    }
+  }
+
+  @Override
+  public Map<String, Object> completeDestinationOAuth(UUID workspaceId,
+                                                      UUID destinationDefinitionId,
+                                                      Map<String, Object> queryParams,
+                                                      String redirectUrl)
+      throws IOException, ConfigNotFoundException {
+    if (queryParams.containsKey("code")) {
+      final String code = (String) queryParams.get("code");
+      final JsonNode oAuthParamConfig = getDestinationOAuthParamConfig(workspaceId, destinationDefinitionId);
+      if (oAuthParamConfig.has("client_id") && oAuthParamConfig.has("client_secret")) {
+        final String clientId = oAuthParamConfig.get("client_id").asText();
+        final String clientSecret = oAuthParamConfig.get("client_secret").asText();
+        return completeOAuthFlow(clientId, clientSecret, code, redirectUrl);
+      } else {
+        throw new IOException("Undefined parameter 'client_id' and 'client_secret' for Google OAuth Flow.");
+      }
+    } else {
+      throw new IOException("Undefined 'code' from consent redirected url.");
+    }
+  }
+
+  private Map<String, Object> completeOAuthFlow(String clientId, String clientSecret, String code, String redirectUrl) throws IOException {
+    final ImmutableMap<String, String> body = new Builder<String, String>()
+        .put("client_id", clientId)
+        .put("client_secret", clientSecret)
+        .put("code", code)
+        .put("grant_type", "authorization_code")
+        .put("redirect_uri", redirectUrl)
+        .build();
+    final HttpRequest request = HttpRequest.newBuilder()
+        .POST(HttpRequest.BodyPublishers.ofString(toUrlEncodedString(body)))
+        .uri(URI.create(GOOGLE_ANALYTICS_ACCESS_TOKEN_URL))
+        .header("Content-Type", "application/x-www-form-urlencoded")
+        .build();
+    final HttpResponse<String> response;
+    try {
+      response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+      var data = Jsons.deserialize(response.body());
+      if (data.has("refresh_token")) {
+        return Map.of("refresh_token", data.get("refresh_token").asText());
```

Comment:
I am thinking we should probably include the `access_token` that we retrieve from Google API from the first `completeOAuthFlow` call (it gives back both access & refresh tokens), WDYT?

The access token may expire and thus, a connector should certainly rely on the refresh token In order to get a new one autonomously. But that means the google connector needs to implement OAuth logic from the start...

To make the code simpler for a new connector, it could just use the access token as provided from airbyte server and start focusing efforts on replicating data instead of oauth.

When the connector faces expiring access token, they'd need to implement oauth logic to use the refresh token as a next step.

Reply:
it's useful to provide the access token to the connector so we don't incur unnecessary token refreshes (google caps those). Unfortunately though, for any oauth-based google connector to be useful, it will _probably_  need to implement refresh logic from the beginning. So it's not a big deal to make that assumption.